include/pybind11/factory_support.h

```diff
@@ -0,0 +1,280 @@
+    template <typename Class, enable_if_t<Class::has_alias, int> = 0>
+    static bool is_alias(Cpp<Class> *ptr) {
+        return dynamic_cast<Alias<Class> *>(ptr) != nullptr;
+    }
+    // Failing fallback version for a no-alias class (always returns false)
+    template <typename Class>
+    constexpr static bool is_alias(void *) { return false; }
+
+    // Attempts to constructs an alias using a `Alias(Cpp &&)` constructor.  Returns true if the
+    // constructor existed, false is the alias cannot be constructed that way.
+    template <typename Class, enable_if_t<std::is_constructible<Alias<Class>, Cpp<Class> &&>::value, int> = 0>
+    static bool construct_alias_from_cpp(value_and_holder &v_h, Cpp<Class> &&base) {
+        new (preallocate(v_h)) Alias<Class>(std::move(base));
+        return true;
+    }
+    // Base version for aliases without an appropriate constructor; does nothing and returns false.
+    template <typename Class, enable_if_t<!std::is_constructible<Alias<Class>, Cpp<Class> &&>::value, int> = 0>
+    static bool construct_alias_from_cpp(value_and_holder &, Cpp<Class> &&) { return false; }
+
+    // Error-generating fallback
+    template <typename Class>
+    static void construct(...) {
+        static_assert(!std::is_same<Class, Class>::value /* always false */,
+                "pybind11::init(): wrapped factory function must return a compatible pointer, "
+                "holder, or value");
+    }
+
+    // Pointer return v1: a factory function that returns a class pointer for a registered class
+    // without an alias
+    template <typename Class, enable_if_t<!Class::has_alias, int> = 0>
+    static void construct(value_and_holder &v_h, Cpp<Class> *ptr, PyTypeObject *) {
+        no_nullptr(ptr);
+        deallocate(v_h) = ptr;
+    }
+
+    // Pointer return v2: a factory that always returns an alias instance ptr (like py::init_alias)
+    template <typename Class, enable_if_t<Class::has_alias, int> = 0>
+    static void construct(value_and_holder &v_h, Alias<Class> *alias_ptr, PyTypeObject *) {
+        no_nullptr(alias_ptr);
+        deallocate(v_h) = static_cast<Cpp<Class> *>(alias_ptr);
+    }
+
+    // Pointer return v3: returning a cpp class for a registered class with an alias.  If we don't
+    // need an alias, we simply use the cpp pointer.  Otherwise, we try a dynamic_cast to see if the
+    // cpp pointer is actually a polymorphic alias instance.  If it isn't, but we can construct the
+    // alias via an `Alias(Cpp &&)` constructor, we do.  Otherwise we throw a type error.
+    template <typename Class, enable_if_t<Class::has_alias, int> = 0>
+    static void construct(value_and_holder &v_h, Cpp<Class> *ptr, PyTypeObject *cl_type) {
+        no_nullptr(ptr);
+        if (Py_TYPE(v_h.inst) != cl_type) {
+            // Inherited from on the Python side: an alias instance is required
+            if (!is_alias<Class>(ptr)) {
```

Comment:
It should have been, but may have been dropped accidentally somewhere along the line.  I'll investigate.